File: src/ipc/catalog.ts
import {
  createInternalError,
  hasControlCharacters,
  invokeCommand,
  isRecord,
} from "./core";

export const CATALOG_SCHEMA_VERSION = 1;
const MAX_MANAGED_APPS = 32;
const MAX_CONFIG_DOCUMENTS_PER_APP = 16;
const MAX_PATH_VARIANTS_PER_DOCUMENT = 8;
const MAX_CONFIG_DOCUMENT_BYTES = 2 * 1024 * 1024;
const MAX_CAPABILITIES = 8;
const MAX_TEXT_BYTES = 512;

export type ManagedAppCapability =
  | "DETECT"
  | "READ_CONFIG"
  | "WRITE_CONFIG"
  | "MANAGE_SERVICE";

export type ManagedAppCoverageClass =
  | "MANAGED_WRITABLE"
  | "MANAGED_READ_ONLY"
  | "DETECTED_UNSUPPORTED"
  | "EXCLUDED";

export type CatalogPriority = "PRIORITY_A" | "PRIORITY_B" | "PRIORITY_C";

export type DetectionEvidenceKind =
  | "BREW_CASK"
  | "BREW_FORMULA"
  | "EXECUTABLE"
  | "HOMEBREW_PATH"
  | "HOME_PATH"
  | "SERVICE";

export type ConfigPathRoot =
  | "HOME"
  | "XDG_CONFIG_HOME"
  | "APPLICATION_SUPPORT"
  | "HOMEBREW_PREFIX"
  | "APP_SUPPORT";

export type ConfigPathExistenceRule =
  | "FILE"
  | "DIRECTORY"
  | "FILE_OR_DIRECTORY";

export type ConfigFormat =
  | "CADDYFILE"
  | "GIT_CONFIG"
  | "INI"
  | "JSON"
  | "JSONC"
  | "KEY_VALUE"
  | "MARKDOWN"
  | "MARKDOWN_DIRECTORY"
  | "PLIST"
  | "SHELL"
  | "SSH_CONFIG"
  | "TEXT"
  | "TOML"
  | "YAML";

export type ConfigFormatFamily =
  | "JSON"
  | "JSONC"
  | "TOML"
  | "YAML"
  | "INI"
  | "GIT_CONFIG"
  | "KEY_VALUE"
  | "PLIST"
  | "COMMAND"
  | "PLAIN_TEXT";

export type ConfigSensitivity = "STANDARD" | "SENSITIVE" | "SECRET";

export type ConfigAccessMode =
  | "READ_WRITE"
  | "READ_ONLY"
  | "METADATA_ONLY"
  | "EXCLUDED";

export interface ManagedConfigPathVariantPresentation {
  variantId: string;
  root: ConfigPathRoot;
  relativePath: string;
  existenceRule: ConfigPathExistenceRule;
  precedence: number;
}

export interface ManagedConfigDocumentPresentation {
  configId: string;
  purpose: string;
  pathVariants: readonly ManagedConfigPathVariantPresentation[];
  format: ConfigFormat;
  formatFamily: ConfigFormatFamily;
  sensitivity: ConfigSensitivity;
  accessMode: ConfigAccessMode;
  editorKey: string;
  maxSizeBytes: number;
}

export interface ManagedAppPresentation {
  category: string;
  configDocuments: readonly ManagedConfigDocumentPresentation[];
}

export interface ManagedAppDetectionEvidence {
  kind: DetectionEvidenceKind;
  value: string;
}

export interface ManagedAppSupportPresentation {
  limitations: readonly string[];
  exclusions: readonly string[];
  requirement: string;
}

export interface ManagedAppSummary {
  id: string;
  displayName: string;
  description: string;
  iconKey: string;
  priority: CatalogPriority;
  coverageClass: ManagedAppCoverageClass;
  presentation: ManagedAppPresentation;
  detectionEvidence: readonly ManagedAppDetectionEvidence[];
  support: ManagedAppSupportPresentation;
  capabilities: readonly ManagedAppCapability[];
  managedDocumentCount: number;
  serviceCount: number;
}

export interface CatalogCoveragePolicy {
  priorityATotal: number;
  priorityAUsable: number;
  priorityBTotal: number;
  priorityBCovered: number;
  minimumEligibleTextPercent: number;
}

export interface ManagedAppCatalog {
  schemaVersion: typeof CATALOG_SCHEMA_VERSION;
  coveragePolicy: CatalogCoveragePolicy;
  applications: readonly ManagedAppSummary[];
}

const CAPABILITIES: readonly string[] = [
  "DETECT",
  "READ_CONFIG",
  "WRITE_CONFIG",
  "MANAGE_SERVICE",
];
const COVERAGE_CLASSES: readonly string[] = [
  "MANAGED_WRITABLE",
  "MANAGED_READ_ONLY",
  "DETECTED_UNSUPPORTED",
  "EXCLUDED",
];
const PRIORITIES: readonly string[] = [
  "PRIORITY_A",
  "PRIORITY_B",
  "PRIORITY_C",
];
const DETECTION_EVIDENCE_KINDS: readonly string[] = [
  "BREW_CASK",
  "BREW_FORMULA",
  "EXECUTABLE",
  "HOMEBREW_PATH",
  "HOME_PATH",
  "SERVICE",
];
const PATH_ROOTS: readonly string[] = [
  "HOME",
  "XDG_CONFIG_HOME",
  "APPLICATION_SUPPORT",
  "HOMEBREW_PREFIX",
  "APP_SUPPORT",
];
const EXISTENCE_RULES: readonly string[] = [
  "FILE",
  "DIRECTORY",
  "FILE_OR_DIRECTORY",
];
export const CONFIG_FORMATS: readonly string[] = [
  "CADDYFILE",
  "GIT_CONFIG",
  "INI",
  "JSON",
  "JSONC",
  "KEY_VALUE",
  "MARKDOWN",
  "MARKDOWN_DIRECTORY",
  "PLIST",
  "SHELL",
  "SSH_CONFIG",
  "TEXT",
  "TOML",
  "YAML",
];
const FORMAT_FAMILIES: readonly string[] = [
  "JSON",
  "JSONC",
  "TOML",
  "YAML",
  "INI",
  "GIT_CONFIG",
  "KEY_VALUE",
  "PLIST",
  "COMMAND",
  "PLAIN_TEXT",
];
const SENSITIVITIES: readonly string[] = [
  "STANDARD",
  "SENSITIVE",
  "SECRET",
];
const ACCESS_MODES: readonly string[] = [
  "READ_WRITE",
  "READ_ONLY",
  "METADATA_ONLY",
  "EXCLUDED",
];
const FORMAT_FAMILY_BY_FORMAT: Readonly<Record<ConfigFormat, ConfigFormatFamily>> = {
  CADDYFILE: "COMMAND",
  GIT_CONFIG: "GIT_CONFIG",
  INI: "INI",
  JSON: "JSON",
  JSONC: "JSONC",
  KEY_VALUE: "KEY_VALUE",
  MARKDOWN: "PLAIN_TEXT",
  MARKDOWN_DIRECTORY: "PLAIN_TEXT",
  PLIST: "PLIST",
  SHELL: "COMMAND",
  SSH_CONFIG: "COMMAND",
  TEXT: "PLAIN_TEXT",
  TOML: "TOML",
  YAML: "YAML",
};

function decodeText(value: unknown): string {
  if (
    typeof value !== "string" ||
    value.trim().length === 0 ||
    new TextEncoder().encode(value).length > MAX_TEXT_BYTES
  ) {
    throw createInternalError();
  }
  return value;
}

function decodeIdentifier(value: unknown): string {
  const identifier = decodeText(value);
  if (!/^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(identifier)) {
    throw createInternalError();
  }
  return identifier;
}

function decodeCount(value: unknown): number {
  if (
    typeof value !== "number" ||
    !Number.isSafeInteger(value) ||
    value < 0 ||
    value > 32
  ) {
    throw createInternalError();
  }
  return value;
}

function decodePercentage(value: unknown): number {
  if (
    typeof value !== "number" ||
    !Number.isSafeInteger(value) ||
    value < 0 ||
    value > 100
  ) {
    throw createInternalError();
  }
  return value;
}

function decodeBoundedSize(value: unknown): number {
  if (
    typeof value !== "number" ||
    !Number.isSafeInteger(value) ||
    value < 1 ||
    value > MAX_CONFIG_DOCUMENT_BYTES
  ) {
    throw createInternalError();
  }
  return value;
}

function decodeKnownValue<T extends string>(
  value: unknown,
  allowed: readonly string[],
): T {
  if (typeof value !== "string" || !allowed.includes(value)) {
    throw createInternalError();
  }
  return value as T;
}

function decodeRelativePath(value: unknown): string {
  const relativePath = decodeText(value);
  const parts = relativePath.split("/");
  if (
    relativePath.startsWith("/") ||
    relativePath.includes("\\") ||
    relativePath.includes("\0") ||
    hasControlCharacters(relativePath) ||
    parts.some((part) => part.length === 0 || part === "." || part === "..")
  ) {
    throw createInternalError();
  }
  return relativePath;
}

function decodePathVariants(
  value: unknown,
): readonly ManagedConfigPathVariantPresentation[] {
  if (
    !Array.isArray(value) ||
    value.length === 0 ||
    value.length > MAX_PATH_VARIANTS_PER_DOCUMENT
  ) {
    throw createInternalError();
  }

  const variants = value.map((variant, index) => {
    if (!isRecord(variant) || variant.precedence !== index) {
      throw createInternalError();
    }
    return {
      variantId: decodeIdentifier(variant.variantId),
      root: decodeKnownValue<ConfigPathRoot>(variant.root, PATH_ROOTS),
      relativePath: decodeRelativePath(variant.relativePath),
      existenceRule: decodeKnownValue<ConfigPathExistenceRule>(
        variant.existenceRule,
        EXISTENCE_RULES,
      ),
      precedence: index,
    };
  });
  if (
    new Set(variants.map((variant) => variant.variantId)).size !==
    variants.length
  ) {
    throw createInternalError();
  }
  return variants;
}

function decodeCapabilities(value: unknown): readonly ManagedAppCapability[] {
  if (
    !Array.isArray(value) ||
    value.length === 0 ||
    value.length > MAX_CAPABILITIES ||
    value.some(
      (capability) =>
        typeof capability !== "string" ||
        !CAPABILITIES.includes(capability),
    )
  ) {
    throw createInternalError();
  }

  const capabilities = value as ManagedAppCapability[];
  if (new Set(capabilities).size !== capabilities.length) {
    throw createInternalError();
  }
  return [...capabilities];
}

function decodeTextList(value: unknown): readonly string[] {
  if (!Array.isArray(value) || value.length === 0 || value.length > 8) {
    throw createInternalError();
  }
  return value.map(decodeText);
}

function decodeDetectionEvidence(
  value: unknown,
): readonly ManagedAppDetectionEvidence[] {
  if (!Array.isArray(value) || value.length === 0 || value.length > 32) {
    throw createInternalError();
  }
  return value.map((evidence) => {
    if (!isRecord(evidence)) {
      throw createInternalError();
    }
    return {
      kind: decodeKnownValue<DetectionEvidenceKind>(
        evidence.kind,
        DETECTION_EVIDENCE_KINDS,
      ),
      value: decodeText(evidence.value),
    };
  });
}

function decodeSupport(value: unknown): ManagedAppSupportPresentation {
  if (!isRecord(value)) {
    throw createInternalError();
  }
  return {
    limitations: decodeTextList(value.limitations),
    exclusions: decodeTextList(value.exclusions),
    requirement: decodeText(value.requirement),
  };
}

function decodeCoverageClass(
  value: unknown,
  capabilities: readonly ManagedAppCapability[],
): ManagedAppCoverageClass {
  const inferred = capabilities.includes("WRITE_CONFIG")
    ? "MANAGED_WRITABLE"
    : capabilities.includes("READ_CONFIG")
      ? "MANAGED_READ_ONLY"
      : "DETECTED_UNSUPPORTED";
  const coverageClass = value === undefined ? inferred : value;
  if (
    typeof coverageClass !== "string" ||
    !COVERAGE_CLASSES.includes(coverageClass)
  ) {
    throw createInternalError();
  }

  const canRead = capabilities.includes("READ_CONFIG");
  const canWrite = capabilities.includes("WRITE_CONFIG");
  if (
    (coverageClass === "MANAGED_WRITABLE" && (!canRead || !canWrite)) ||
    (coverageClass === "MANAGED_READ_ONLY" && (!canRead || canWrite)) ||
    ((coverageClass === "DETECTED_UNSUPPORTED" ||
      coverageClass === "EXCLUDED") &&
      (canRead || canWrite))
  ) {
    throw createInternalError();
  }
  return coverageClass as ManagedAppCoverageClass;
}

function decodePresentation(
  value: unknown,
  managedDocumentCount: number,
): ManagedAppPresentation {
  if (value === undefined) {
    return { category: "Other", configDocuments: [] };
  }
  if (
    !isRecord(value) ||
    !Array.isArray(value.configDocuments) ||
    value.configDocuments.length > MAX_CONFIG_DOCUMENTS_PER_APP
  ) {
    throw createInternalError();
  }

  const configDocuments = value.configDocuments.map((document) => {
    if (!isRecord(document)) {
      throw createInternalError();
    }
    const format = decodeKnownValue<ConfigFormat>(
      document.format,
      CONFIG_FORMATS,
    );
    const formatFamily = decodeKnownValue<ConfigFormatFamily>(
      document.formatFamily,
      FORMAT_FAMILIES,
    );
    if (FORMAT_FAMILY_BY_FORMAT[format] !== formatFamily) {
      throw createInternalError();
    }
    return {
      configId: decodeIdentifier(document.configId),
      purpose: decodeText(document.purpose),
      pathVariants: decodePathVariants(document.pathVariants),
      format,
      formatFamily,
      sensitivity: decodeKnownValue<ConfigSensitivity>(
        document.sensitivity,
        SENSITIVITIES,
      ),
      accessMode: decodeKnownValue<ConfigAccessMode>(
        document.accessMode,
        ACCESS_MODES,
      ),
      editorKey: decodeIdentifier(document.editorKey),
      maxSizeBytes: decodeBoundedSize(document.maxSizeBytes),
    };
  });
  if (
    new Set(configDocuments.map((document) => document.configId)).size !==
      configDocuments.length ||
    configDocuments.length !== managedDocumentCount
  ) {
    throw createInternalError();
  }

  return {
    category: decodeText(value.category),
    configDocuments,
  };
}

function decodeManagedApp(value: unknown): ManagedAppSummary {
  if (!isRecord(value)) {
    throw createInternalError();
  }

  const capabilities = decodeCapabilities(value.capabilities);
  const managedDocumentCount = decodeCount(value.managedDocumentCount);
  return {
    id: decodeIdentifier(value.id),
    displayName: decodeText(value.displayName),
    description: decodeText(value.description),
    iconKey: decodeIdentifier(value.iconKey),
    priority: decodeKnownValue<CatalogPriority>(value.priority, PRIORITIES),
    coverageClass: decodeCoverageClass(value.coverageClass, capabilities),
    presentation: decodePresentation(value.presentation, managedDocumentCount),
    detectionEvidence: decodeDetectionEvidence(value.detectionEvidence),
    support: decodeSupport(value.support),
    capabilities,
    managedDocumentCount,
    serviceCount: decodeCount(value.serviceCount),
  };
}

function decodeCoveragePolicy(value: unknown): CatalogCoveragePolicy {
  if (!isRecord(value)) {
    throw createInternalError();
  }
  const priorityATotal = decodeCount(value.priorityATotal);
  const priorityAUsable = decodeCount(value.priorityAUsable);
  const priorityBTotal = decodeCount(value.priorityBTotal);
  const priorityBCovered = decodeCount(value.priorityBCovered);
  if (
    priorityAUsable > priorityATotal ||
    priorityBCovered > priorityBTotal
  ) {
    throw createInternalError();
  }
  return {
    priorityATotal,
    priorityAUsable,
    priorityBTotal,
    priorityBCovered,
    minimumEligibleTextPercent: decodePercentage(
      value.minimumEligibleTextPercent,
    ),
  };
}

export function decodeManagedAppCatalog(value: unknown): ManagedAppCatalog {
  if (
    !isRecord(value) ||
    value.schemaVersion !== CATALOG_SCHEMA_VERSION ||
    !Array.isArray(value.applications) ||
    value.applications.length === 0 ||
    value.applications.length > MAX_MANAGED_APPS
  ) {
    throw createInternalError();
  }

  const applications = value.applications.map(decodeManagedApp);
  if (new Set(applications.map((app) => app.id)).size !== applications.length) {
    throw createInternalError();
  }
  const coveragePolicy = decodeCoveragePolicy(value.coveragePolicy);
  const priorityAApplications = applications.filter(
    (application) => application.priority === "PRIORITY_A",
  );
  const priorityBApplications = applications.filter(
    (application) => application.priority === "PRIORITY_B",
  );
  const priorityAUsable = priorityAApplications.filter(
    (application) =>
      application.coverageClass === "MANAGED_WRITABLE" &&
      application.capabilities.includes("READ_CONFIG") &&
      application.capabilities.includes("WRITE_CONFIG"),
  ).length;
  const priorityBCovered = priorityBApplications.filter(
    (application) =>
      application.coverageClass === "MANAGED_WRITABLE" ||
      application.coverageClass === "MANAGED_READ_ONLY" ||
      application.coverageClass === "EXCLUDED",
  ).length;
  if (
    priorityAApplications.length !== coveragePolicy.priorityATotal ||
    priorityAUsable !== coveragePolicy.priorityAUsable ||
    priorityBApplications.length !== coveragePolicy.priorityBTotal ||
    priorityBCovered !== coveragePolicy.priorityBCovered
  ) {
    throw createInternalError();
  }

  return {
    schemaVersion: CATALOG_SCHEMA_VERSION,
    coveragePolicy,
    applications,
  };
}

export function listManagedApps(): Promise<ManagedAppCatalog> {
  return invokeCommand("list_managed_apps", decodeManagedAppCatalog);
}

File: src-tauri/src/commands/catalog.rs
use serde::Serialize;

use crate::{
    catalog::{
        CATALOG_SCHEMA_VERSION, CatalogCoverageClass, CatalogPriority, ConfigAccessMode,
        ConfigFormatFamily, ConfigPathExistenceRule, ConfigSensitivity,
        MINIMUM_ELIGIBLE_TEXT_COVERAGE_PERCENT, ManagedAppDefinition, PRIORITY_A_APP_IDS,
        PRIORITY_B_APP_IDS, catalog_priority, load_builtin_catalog,
    },
    error::AppError,
    security::paths::CatalogPathRoot,
};

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ManagedConfigPathVariantPresentation {
    variant_id: String,
    root: CatalogPathRoot,
    relative_path: String,
    existence_rule: ConfigPathExistenceRule,
    precedence: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ManagedConfigDocumentPresentation {
    config_id: String,
    purpose: String,
    path_variants: Vec<ManagedConfigPathVariantPresentation>,
    format: String,
    format_family: ConfigFormatFamily,
    sensitivity: ConfigSensitivity,
    access_mode: ConfigAccessMode,
    editor_key: String,
    max_size_bytes: usize,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ManagedAppPresentation {
    category: String,
    config_documents: Vec<ManagedConfigDocumentPresentation>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DetectionEvidencePresentation {
    kind: String,
    value: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ManagedAppSupportPresentation {
    limitations: Vec<String>,
    exclusions: Vec<String>,
    requirement: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ManagedAppSummary {
    id: String,
    display_name: String,
    description: String,
    icon_key: String,
    priority: CatalogPriority,
    coverage_class: CatalogCoverageClass,
    presentation: ManagedAppPresentation,
    detection_evidence: Vec<DetectionEvidencePresentation>,
    support: ManagedAppSupportPresentation,
    capabilities: Vec<String>,
    managed_document_count: usize,
    service_count: usize,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CatalogCoveragePolicy {
    priority_a_total: usize,
    priority_a_usable: usize,
    priority_b_total: usize,
    priority_b_covered: usize,
    minimum_eligible_text_percent: u8,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ManagedAppCatalog {
    schema_version: u16,
    coverage_policy: CatalogCoveragePolicy,
    applications: Vec<ManagedAppSummary>,
}

#[tauri::command]
pub fn list_managed_apps() -> Result<ManagedAppCatalog, AppError> {
    let catalog = load_builtin_catalog().map_err(|_| AppError::internal())?;
    let applications = catalog
        .apps()
        .iter()
        .map(|app| ManagedAppSummary {
            id: app.id().to_owned(),
            display_name: app.display_name().to_owned(),
            description: app.description().to_owned(),
            icon_key: app.icon_key().to_owned(),
            priority: catalog_priority(app.id()),
            coverage_class: app.coverage_class(),
            presentation: ManagedAppPresentation {
                category: app.presentation_category().to_owned(),
                config_documents: app
                    .config_documents()
                    .iter()
                    .map(|document| ManagedConfigDocumentPresentation {
                        config_id: document.config_id().to_owned(),
                        purpose: document.purpose().to_owned(),
                        path_variants: document
                            .path_variants()
                            .iter()
                            .map(|variant| ManagedConfigPathVariantPresentation {
                                variant_id: variant.variant_id().to_owned(),
                                root: variant.root(),
                                relative_path: variant.relative_path().to_owned(),
                                existence_rule: variant.existence_rule(),
                                precedence: variant.precedence(),
                            })
                            .collect(),
                        format: document.format().to_owned(),
                        format_family: document.format_family(),
                        sensitivity: document.sensitivity_kind(),
                        access_mode: document.access_mode(),
                        editor_key: document.editor_key().to_owned(),
                        max_size_bytes: document.max_size_bytes(),
                    })
                    .collect(),
            },
            detection_evidence: app
                .detection_rules()
                .iter()
                .map(|rule| DetectionEvidencePresentation {
                    kind: rule.kind().to_owned(),
                    value: safe_detection_value(rule.kind(), rule.value()),
                })
                .collect(),
            support: support_presentation(app),
            capabilities: app.capabilities().to_vec(),
            managed_document_count: app.config_documents().len(),
            service_count: app.services().len(),
        })
        .collect();

    Ok(ManagedAppCatalog {
        schema_version: CATALOG_SCHEMA_VERSION,
        coverage_policy: CatalogCoveragePolicy {
            priority_a_total: PRIORITY_A_APP_IDS.len(),
            priority_a_usable: catalog
                .apps()
                .iter()
                .filter(|app| {
                    catalog_priority(app.id()) == CatalogPriority::PriorityA
                        && app.coverage_class() == CatalogCoverageClass::ManagedWritable
                })
                .count(),
            priority_b_total: PRIORITY_B_APP_IDS.len(),
            priority_b_covered: catalog
                .apps()
                .iter()
                .filter(|app| {
                    catalog_priority(app.id()) == CatalogPriority::PriorityB
                        && matches!(
                            app.coverage_class(),
                            CatalogCoverageClass::ManagedWritable
                                | CatalogCoverageClass::ManagedReadOnly
                                | CatalogCoverageClass::Excluded
                        )
                })
                .count(),
            minimum_eligible_text_percent: MINIMUM_ELIGIBLE_TEXT_COVERAGE_PERCENT,
        },
        applications,
    })
}

fn safe_detection_value(kind: &str, value: &str) -> String {
    if kind == "HOME_PATH" {
        if let Some(relative) = value.strip_prefix("~/Library/Application Support/") {
            return format!("APPLICATION_SUPPORT/{relative}");
        }
        if let Some(relative) = value.strip_prefix("~/.config/") {
            return format!("XDG_CONFIG_HOME/{relative}");
        }
        if let Some(relative) = value.strip_prefix("~/") {
            return format!("HOME/{relative}");
        }
    }
    if kind == "HOMEBREW_PATH"
        && let Some(relative) = value.strip_prefix("${HOMEBREW_PREFIX}/")
    {
        return format!("HOMEBREW_PREFIX/{relative}");
    }
    value.to_owned()
}

fn support_presentation(app: &ManagedAppDefinition) -> ManagedAppSupportPresentation {
    let limitation = match app.coverage_class() {
        CatalogCoverageClass::ManagedWritable => "仅支持 catalog 明确授权的配置文档与操作。",
        CatalogCoverageClass::ManagedReadOnly => {
            "仅支持有界读取；不提供写入、备份、恢复、服务或提权操作。"
        }
        CatalogCoverageClass::DetectedUnsupported => {
            "仅检测存在性；不读取配置内容，也不执行管理操作。"
        }
        CatalogCoverageClass::Excluded => {
            "仅展示排除原因与检测证据；不读取内容，也不执行管理操作。"
        }
    };
    let requirement = app.support_requirement().unwrap_or(match app.coverage_class() {
        CatalogCoverageClass::ManagedWritable => {
            "新增路径或格式仍需独立 catalog 授权与安全审查。"
        }
        CatalogCoverageClass::ManagedReadOnly => {
            "提升为可写需要批准 parser、round-trip、validator、sensitivity policy 与 editor capability。"
        }
        CatalogCoverageClass::DetectedUnsupported => {
            "需要明确的设置路径、格式、解析器或产品契约后才能增加支持。"
        }
        CatalogCoverageClass::Excluded => {
            "只有在设置可与敏感或运行时数据可靠分离后才能重新评估支持。"
        }
    });

    ManagedAppSupportPresentation {
        limitations: vec![limitation.to_owned(), app.description().to_owned()],
        exclusions: vec![
            "未在 configDocuments 中声明的路径不获授权。".to_owned(),
            "凭据、会话、历史、日志、缓存、数据库、遥测和运行时状态仅保留分类元数据。".to_owned(),
        ],
        requirement: requirement.to_owned(),
    }
}
